python/config_settings/transition.bzl

```diff
@@ -0,0 +1,200 @@
+"""The transition module contains the rule definitions to wrap py_binary and py_test and transition
+them to the desired target platform.
+"""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//python:defs.bzl", _py_binary = "py_binary", _py_test = "py_test")
+
+def _transition_python_version_impl(_, attr):
+    return {"//python/config_settings:python_version": str(attr.python_version)}
+
+_transition_python_version = transition(
+    implementation = _transition_python_version_impl,
+    inputs = [],
+    outputs = ["//python/config_settings:python_version"],
+)
+
+def _transition_py_impl(ctx):
+    target = ctx.attr.target[0]
+    executable = ctx.actions.declare_file(ctx.attr.name + (".exe" if ctx.attr.is_windows else ""))
+    ctx.actions.symlink(
+        is_executable = True,
+        output = executable,
+        target_file = target[DefaultInfo].files_to_run.executable,
+    )
+    zipfile_symlink = None
+    if ctx.attr.is_windows:
+        zipfile = None
+        expected_target_path = target[DefaultInfo].files_to_run.executable.short_path[:-4] + ".zip"
+        for file in target[DefaultInfo].default_runfiles.files.to_list():
+            if file.short_path == expected_target_path:
+                zipfile = file
+        zipfile_symlink = ctx.actions.declare_file(ctx.attr.name + ".zip")
+        ctx.actions.symlink(
+            is_executable = True,
+            output = zipfile_symlink,
+            target_file = zipfile,
+        )
+    env = {}
+    for k, v in ctx.attr.env.items():
+        env[k] = ctx.expand_location(v)
+    providers = [
+        DefaultInfo(
+            executable = executable,
+            files = depset([zipfile_symlink] if zipfile_symlink else [], transitive = [target[DefaultInfo].files]),
+            runfiles = ctx.runfiles([zipfile_symlink] if zipfile_symlink else []).merge(target[DefaultInfo].default_runfiles),
+        ),
+        target[PyInfo],
+        target[PyRuntimeInfo],
+        target[InstrumentedFilesInfo],
+        target[OutputGroupInfo],
+        # TODO(f0rmiga): testing.TestEnvironment is deprecated in favour of RunEnvironmentInfo but
+        # RunEnvironmentInfo is not exposed in Bazel < 5.3.
+        # https://github.com/bazelbuild/bazel/commit/dbdfa07e92f99497be9c14265611ad2920161483
+        testing.TestEnvironment(environment = env),
+    ]
+    return providers
+
+_COMMON_ATTRS = {
+    "deps": attr.label_list(
+        mandatory = False,
+    ),
+    "env": attr.string_dict(
+        mandatory = False,
+    ),
+    "is_windows": attr.bool(
+        mandatory = True,
+    ),
+    "python_version": attr.string(
+        mandatory = True,
+    ),
+    "srcs": attr.label_list(
+        allow_files = True,
+        mandatory = False,
+    ),
+    "target": attr.label(
+        executable = True,
+        cfg = _transition_python_version,
```

Comment:
You'll want to apply the transition to the whole rule (`rule(cfg=...)`, otherwise things like deps, srcs, and tools won't select based upon the same version of python as what's in the `target` attribute.

Reply:
Addressed in 9c833f1.